### PR TITLE
Add a minHeight block support under dimensions

### PR DESCRIPTION
**Pull request by andrewserong**
## Dependencies

* [x] https://github.com/WordPress/gutenberg/pull/45334

## What?

Add a `minHeight` block support feature that sits under Dimensions in the block editor inspector and in global styles.

This PR builds on the good earlier work by @aaronrobertshaw in #32499 which looked at adding in a height block support feature, and the work by @ramonjd on min-height block support in #33970.

## Why?

I think we might be at a point where adding min-height will be really useful, so let's revisit it now that we've got the style engine in 😄

Minimum height is quite useful for defining the minimum vertical dimensions of a block, while allowing it to expand beyond that size. As raised in https://github.com/WordPress/gutenberg/issues/45117 having control over minimum height could be particularly useful for blocks like Post Content when used in page layouts, and in the Group block and its Row and Stack variations.

## Issues that might be resolved or partially resolved by including this block support

* https://github.com/WordPress/gutenberg/issues/28845
* https://github.com/WordPress/gutenberg/issues/45117

## How?

* This PR mostly includes or builds on changes in the earlier preceding PRs (kudos Aaron and Ramon), but uses the style engine now.
* Add a `dimensions` key to the block supports and to the style object in block attributes, and a `minHeight` property. (I avoided the `__experimental` prefix for now, since we've been pretty settled on `dimensions` in earlier explorations, but can re-instate it if folks prefer)
* For now, use a simple `UnitControl` for the minimum height input
* Include `minHeight` in the list of controls enabled by the appearance tools opt-in

## Testing Instructions

1. In a post, add a Group block with a paragraph or two within it.
2. From the dimensions panel click to expose the `Minimum height` control and test that it works as expected.
3. In global styles, try adjusting the Group block's minimum height at the block level.
4. In a single or page template, try adjusting the minimum height of the Post Content block — it should result in being able to set a minimum height — this will likely be useful to use with `vh` units.

## Screenshots or screencast 

| Post Content block | Group block |
| --- | --- |
| image | image |

### CODE REVIEW

File: packages/style-engine/class-wp-style-engine.php
Comment (ramonjd) [inline packages/style-engine/class-wp-style-engine.php]: This is a good top-level label. 

Carry on.

Comment (ramonjd) [reply packages/style-engine/class-wp-style-engine.php]: Side note: I'm working on a PR that will allow us to extend base style definitions

- https://github.com/WordPress/gutenberg/pull/45296

Nothing to do here. I think having a set of definitions for `dimensions: { width, height, ... }` in the base set of definitions makes sense 👍

Comment (ramonjd) [reply packages/style-engine/class-wp-style-engine.php]: When the time comes, after testing this branch and so on, what do you think about splitting the style engine additions into a dependency PR? 

I'm just thinking of the need for unit tests and how to avoid bloating this PR.

Comment (andrewserong) [reply packages/style-engine/class-wp-style-engine.php]: > what do you think about splitting the style engine additions into a dependency PR?

Thanks for taking a look!

Absolutely, I'm all for splitting this up into component parts to make the merge and review process easier. It's super easy for these new block support addition PRs to become unwieldy 👍

**Comment by ramonjd** [comment]
This is testing well manually. 

https://user-images.githubusercontent.com/6458278/198187783-9fa0d93b-0c64-4289-9247-b639950fd265.mp4


Love the fact that I can add `100vh` to a template to force full height


https://user-images.githubusercontent.com/6458278/198187768-5dadbc74-0325-4ac5-8029-137c003479cc.mp4

File: lib/compat/wordpress-6.2/class-wp-theme-json-6-2.php
Comment (aaronrobertshaw) [inline lib/compat/wordpress-6.2/class-wp-theme-json-6-2.php]: Extreme OCD nit: Should we maintain the alphabetical ordering for the rest of this array? Keep the sorting of the other config arrays going as well?

File: packages/block-editor/src/hooks/style.js
Comment (aaronrobertshaw) [inline packages/block-editor/src/hooks/style.js]: I think we might have a tiny copy-and-paste error here, shouldn't the dimensions skip serialization check look at:
```
[ `${ DIMENSIONS_SUPPORT_KEY }.__experimentalSkipSerialization` ]
```

The edits to block.json files to opt into the min-height is under `dimensions`, so I'd expect to be able to add the skip serialization flag under there as per other supports.

File: packages/style-engine/class-wp-style-engine.php
Comment (aaronrobertshaw) [reply packages/style-engine/class-wp-style-engine.php]: Splitting this up a bit sounds like a solid plan. Might make tracking down when individual blocks had their bespoke controls moved over to this block support easier as well.

**Review by aaronrobertshaw — COMMENTED** [review]
Awesome work @andrewserong, nicely done 👍 

Overall this tests very well. 

I did find one small issue around not being able to skip serialization though and left an inline comment on that.

File: packages/block-editor/src/hooks/style.js
Comment (andrewserong) [reply packages/block-editor/src/hooks/style.js]: Oh, good catch! Went too quickly with the copy+paste 😄

File: lib/compat/wordpress-6.2/class-wp-theme-json-6-2.php
Comment (andrewserong) [reply lib/compat/wordpress-6.2/class-wp-theme-json-6-2.php]: I love a good nit like this, cheers, I've updated the order 👍

**Comment by andrewserong** [comment]
Thanks for reviewing and testing @ramonjd and @aaronrobertshaw!

I'll make a start splitting out the style engine changes into a separate PR as I think that'll be the logical thing to land first (the ability to output `min-height`). Will update here as I go 👍

**Comment by andrewserong** [comment]
Update: I've split out the style engine changes and opened up a separate PR for just that part (with tests) over in https://github.com/WordPress/gutenberg/pull/45334

**Comment by andrewserong** [comment]
Update: here's how the feature currently looks in the post editor:



@WordPress/gutenberg-design does this look like an acceptable MVP for adding in a `min-height` block support feature? To recap:

* This PR seeks to add in a minimum height block support that sits under Dimensions and is hidden by default. It behaves in a similar way to the control in the cover block, but can be used on the Group block and its Row and Stack variations, and the Post Content block
* To begin with, it uses a simple UnitControl component as the control and doesn't yet add in the resizeable box — the idea being that we would implement a simple control to begin with, and can look at more complex interactions in follow-ups
* Minimum height will be a useful setting to be used in conjunction with sticky positioning (being explored separately in https://github.com/WordPress/gutenberg/pull/44723) — for example, imagine being able to set a minimum height on a sticky or fixed heading or footer
* Being able to set minimum height will also be useful when go to add in vertical alignment controls for the Stack block (with a height set, the align controls will be more visibly useful)

My goal for this PR is to try to get to a good MVP for the feature. We can then land it in smaller more easily reviewed PRs (beginning with https://github.com/WordPress/gutenberg/pull/45334), and once the MVP of the feature is in, we can then explore more UI follow-ups like adding in a slider next to the UI control / exploring whether a resizeable box is suitable, too.

_Note: I'm AFK on and off this week and next, so no rush on this one!_

**Comment by aaronrobertshaw** [comment]
> To begin with, it uses a simple UnitControl component as the control and doesn't yet add in the resizeable box — the idea being that we would implement a simple control to begin with, and can look at more complex interactions in follow-ups

When it comes time to address the resizable box side of this new support, one option may be to leverage a block popover for the resizable box. An example can be found in https://github.com/WordPress/gutenberg/pull/41153 where I explored exactly this for the Cover block's min-height. While that PR is a little stale now, it only needs the block popover to have the ability not to be constrained to the viewport to be ready for review.

**Comment by jasmussen** [comment]
Nice, this is cool. I'm assuming we can separately refactor Cover to use this?

Screenshot 2022-10-31 at 12 45 56

And can we call it "Min. height" and add a slider to the mix?

Dimensions

Same componentry as font size, in other words:

Screenshot 2022-10-31 at 12 49 50

Nice enhancement.

**Comment by beafialho** [comment]
Thank you for your work on this! I think this could definitely help solve the [original issue](https://github.com/WordPress/gutenberg/issues/45117).

One important question that arises is if we set a specific minimum height (700px for instance) on desktop, would that adjust for mobile? Where naturally 700px height would be too much.

**Comment by paaljoachim** [comment]
Very nice!

Testing using WP 6.0.3
Twenty Twenty Two
The PR Gutenberg build.

In a page. I added the Minimum Height with 100vh.
Screenshot 2022-10-31 at 13 55 45

Checked the frontend. It looks good!

This is very helpful to have this feature added to the dimensions panel! Thank you!

**Comment by jasmussen** [comment]
One note: while it can be a default control in the Cover block, it shouldn't be a default control in Group or probably any other blocks. It's a useful control when you need it, and it's nice to then add it from the menu. But for the responsive reasons Bea mentions as well, it's not an easy control to use, requiring in many cases an understanding of how `vh` units work. In that light, I see it mostly as a tool for theme and pattern developers, and advanced users, to leverage.

**Comment by andrewserong** [comment]
Thanks for the reviews and feedback everybody! I've rebased this now that #45334 has been merged, so the change here should now be everything left for an MVP of the feature.

> I'm assuming we can separately refactor Cover to use this?

Yes, once we've implemented all the features that exist in the Cover block (e.g. the resizeable box), in a follow-up we should be able to swap out the Cover block's ad hoc control with this block support feature.

> One important question that arises is if we set a specific minimum height (700px for instance) on desktop, would that adjust for mobile? Where naturally 700px height would be too much.

That's a great question. While we don't (yet) have a UI control for editing `clamp()` or `min()` values, it's possible for theme developers (or pattern creators) to manually specify a rule in code. For example, to set `700px` as the min height, but use `100vh` on mobile or viewports smaller than `700px`, then a theme could use the following rule in `theme.json` under `style.blocks`:

```
			"core/post-content": {
				"dimensions": {
					"minHeight": "min(700px, 100vh)"
				}
			},
```

As Joen mentions, I think the complexity of dealing with `vh` units etc is a good argument for keeping the min height control hidden by default, as it currently is in this PR 👍

> And can we call it "Min. height" and add a slider to the mix?

I've updated the labels to `Min. height`. It'd be great to add in the slider, too... Since this PR is already quite large, I was wondering what you think @jasmussen if we looked to land this with the simple `UnitControl` to begin with, and then add in the slider in a follow-up PR? I'm mostly conscious of trying to contain the scope of the PR for reviewing and testing purposes.

If we're happy landing this with a simple UnitControl in the shorter-term, there may be a bit of a delay in me getting the follow-up PR up, though, as I'll be AFK off and on over the next couple of weeks. If it's a blocker for the PR, I'm happy to revisit this one once I'm back properly 🙂

**Review by tellthemachines — COMMENTED** [review]
Just some quick testing, haven't gone through the code in depth, but it seems to be working pretty well! I was curious about combining it with #45364 so did a little experiment adding `fill` styles to a child of a Stack block:

Screenshot 2022-11-02 at 4 48 59 pm

Another thing this PR enables is allowing `space-between` justification for Stack blocks, though we'll have to think about what the controls for that will look like, because "justification" on Stack does horizontal, not vertical justification 🤔 

Anyway, excited for this! I'll look more closely at the code tomorrow.

**Comment by jasmussen** [comment]
> add in the slider in a follow-up PR? I'm mostly conscious of trying to contain the scope of the PR for reviewing and testing purposes.

Seems fine to me so long as we remember to do it 👍 👍

File: packages/block-editor/src/hooks/dimensions.js
Comment (aaronrobertshaw) [inline packages/block-editor/src/hooks/dimensions.js]: I get the feeling that min-height shouldn't really be more prominent than the common spacing controls in this panel such as padding and margin. Would it be better moved lower so toggling it on doesn't stick it above everything else?

Screenshot 2022-11-03 at 10 54 13 am

**Review by aaronrobertshaw — APPROVED** [review]
Nice work getting min height support into shape @andrewserong ✨ 

As others have noted this is testing well for me.

✅ Works for individual blocks
✅ Theme.json and global styles work
✅ Unit tests are passing
✅ Code changes make sense

I did have one question around the placement of the min-height control within the Dimensions panel. I think its position should be moved to have lower prominence when it's toggled on. It might also limit a little the visual movement of controls that a commonly already displayed by default e.g. padding/margin. 

Happy to leave it to you as to whether that happens in a follow-up or not.

**Comment by andrewserong** [comment]
Thanks for all the testing and follow-up everyone, and for the detailed review @aaronrobertshaw! 🙇 

Good call about changing the order of where the min height control displays within the dimensions panel. I've moved it to the bottom of the panel, where it sits much better now: 

image

Once the CI jobs pass, I'll merge this in. I'll then open up another issue to track the follow-up tasks so that we don't forget or miss any of the good ideas and suggestions here.

**Comment by andrewserong** [comment]
I've opened up a follow-up issue to keep track of the tasks and suggestions here: https://github.com/WordPress/gutenberg/issues/45501 — feel free to add anything over there if anyone has any other tasks they think should be addressed!

While it's very early for 6.2, I've added a Needs Dev Note label to this one so we don't forget when the time comes for the 6.2 release, as it'll likely need a little highlighting.

**Comment by andrewserong** [comment]
Update: I've opened up a follow-up PR to add in the slider alongside the unit control: https://github.com/WordPress/gutenberg/pull/45875 — it's the end of my week now, so I very well might have missed something in that PR, but I'll test it some more on Monday.

**Comment by oandregal** [comment]
I wanted to share that https://github.com/WordPress/gutenberg/pull/46102 introduced a regression in the site editor that https://github.com/WordPress/gutenberg/pull/46810 fixes, affecting this feature.

**Comment by andrewserong** [comment]
Thanks for the heads-up, glad to hear it's been resolved! 🙂

**Comment by andrewserong** [comment]
## Dev notes for 6.2

Note: if there's room, this dev note could potentially be combined with other design tools in a larger dev note. Draft below:

---

### Minimum Height dimensions block support

In WordPress 6.2, a new minimum height block support feature has been added, with the Group and Post Content blocks opted-in by default. For themes using the `appearanceTools` feature in `theme.json`, the control will be available in the inspector controls, under Dimensions. It is not exposed by default, and users can access the control via the Dimensions panel's menu.

The feature allows users to set a Group or Post Content block to have a minimum height, similar to the control that exists for the Cover block. An example use case is to ensure that a post content area has a minimum height, forcing a footer area to render lower on the page, even if a page has little content. For this reason, it can be quite useful to use viewport relative units when setting minimum height, e.g. `100vh` for a block that takes up the entire height of the viewport.

Setting minimum height for the Stack variation of the Group block is also useful when combined with the new vertical alignment tools added in 6.2 introduced in [this Gutenberg PR](https://github.com/WordPress/gutenberg/pull/47134), as it allows users to align the blocks within the minimum height of the Stack to be at the `top`, `center`, `bottom` positions within the block, or for the blocks to be aligned via `space-between`.

### How to add minHeight support to a theme

There are two ways to add support for `minHeight` to a block theme. The simplest is to opt in to the `appearanceTools` setting, which automatically enables a number of design tools ([read more in the developer handbook](https://developer.wordpress.org/block-editor/how-to-guides/themes/theme-json/#opt-in-into-ui-controls)).

For themes that wish to have more granular control over which UI tools are enabled, the `minHeight` dimensions support can be opted into by setting `settings.dimensions.minHeight` to `true` in `theme.json`. For example:

```
{
	"settings": {
		"dimensions": {
			"minHeight": true
```

For styling blocks globally, `minHeight` can also be set either within the global styles interface, or within `theme.json`. For example, the Post Content block can be set to have a minimum height of `70vh` by setting `styles.blocks.core/post-content.dimensions.minHeight` to `70vh` in `theme.json`:

```
{
	"styles": {
		"blocks": {
			"core/post-content": {
				"dimensions": {
					"minHeight": "70vh"
				}
```

### How to add minHeight support to a custom block

In the block's `block.json` file, add the `supports.dimensions.minHeight` property, and set it to `true` to enable the minimum height control. For static blocks, the `min-height` CSS rule will be saved as an inline style to post content. For dynamic blocks, the `min-height` rule will be injected into the `style` attribute returned as part of a call to `get_block_wrapper_attributes`. For more information on outputting block supports in dynamic blocks, read [the entry in the developer handbook](https://developer.wordpress.org/block-editor/how-to-guides/block-tutorial/block-supports-in-dynamic-blocks/).

#### Further reading

* The minumum height feature is one part of providing tools to allow sites to align footers to the bottom of the page when the page has no/little content ([Gutenberg issue](https://github.com/WordPress/gutenberg/issues/45117))
* Tracking issue for follow-up tasks beyond 6.2: https://github.com/WordPress/gutenberg/issues/45501
* Block Supports in Dynamic Blocks: https://developer.wordpress.org/block-editor/how-to-guides/block-tutorial/block-supports-in-dynamic-blocks/
* MDN article on relative units: https://developer.mozilla.org/en-US/docs/Learn/CSS/Building_blocks/Values_and_units#relative_length_units

---

@bph for visibility, here's a draft dev note for the minHeight dimensions block support. It could potentially be grouped together with other design tools (e.g. position support?) in a bigger dev note.

**Comment by bph** [comment]
@andrewserong thank you so much for taking care of this dev note. I am leaning towards single post dev note for visibility in the Fieldguide and because it adds so much more flexibility for themes. Would you have bandwidth to add it to the Make blog as a draft, so it can be added to the publishing queue in time for the Fieldguide deadline (later this week)?  LMK, if not, and I can do it for you.

**Comment by andrewserong** [comment]
Thanks @bph, here's a draft on the Make blog for minHeight: https://make.wordpress.org/core/?p=103001&preview=1&_ppp=6452826240

**Comment by merijnponzo** [comment]
Awesome work, will minHeight will be supported for the image block as well?

**Pull request merged**
